File: Boat/GPS/GPS.py
import math
import csv
import serial
import time
import string
import pynmea2


class GPSDataLogger:
    def __init__(self, port="/dev/ttyS0"):
        """
        Initialize the GPSDataLogger.

        Args:
            port (str): Port for GPS device.
        """
        self.port = port

    def parse_latitude(self, data):
        """
        Parse latitude from the NMEA data.

        Args:
            data (str): NMEA data string.

        Returns:
            float: Latitude in decimal degrees.
        """
        lat_data = data.split(',')[3:5]
        latitude = float(lat_data[0][0:2]) + float(lat_data[0][2:]) / 60
        if lat_data[1] == 'S':
            latitude = -latitude
        return latitude

    def parse_longitude(self, data):
        """
        Parse longitude from the NMEA data.

        Args:
            data (str): NMEA data string.

        Returns:
            float: Longitude in decimal degrees.
        """
        lon_data = data.split(',')[5:7]
        longitude = float(lon_data[0][0:3]) + float(lon_data[0][3:]) / 60
        if lon_data[1] == 'W':
            longitude = -longitude
        return longitude

    def parse_time(self, data):
        """
        Parse time from the NMEA data.

        Args:
            data (str): NMEA data string.

        Returns:
            tuple: Hour, minute, and second.
        """
        time_data = data.split(',')[1]
        hour = int(time_data[0:2])
        minute = int(time_data[2:4])
        second = int(time_data[4:6])
        return hour, minute, second

    def parse_date(self, data):
        """
        Parse date from the NMEA data.

        Args:
            data (str): NMEA data string.

        Returns:
            tuple: Day, month, and year.
        """
        date_data = data.split(',')[9]
        day = int(date_data[0:2])
        month = int(date_data[2:4])
        year = int(date_data[4:6])
        return day, month, year

    def log_gps_data(self):
        """
        Continuously receive and log GPS data.
        """
        with open(self.csv_filename, 'w', newline='') as csvfile:
            # Define the fieldnames for the CSV file.
            fieldnames = ['Date', 'Time', 'Latitude', 'Longitude']

            # Create a CSV writer object.
            writer = csv.DictWriter(csvfile, fieldnames=fieldnames)

            # Write the header row to the CSV file.
            writer.writeheader()

            while True:
                # Receive NMEA data.
                ser = serial.Serial(self.port, baudrate=9600, timeout=0.5)
                dataout = pynmea2.NMEAStreamReader()
                nmea_sentence_bytes = ser.readline()
                nmea_sentence_bytes = nmea_sentence_bytes.decode('utf-8')

                if '$GPRMC' in nmea_sentence_bytes:
                    try:
                        # Parse latitude, longitude, time, and date.
                        latitude = round(self.parse_latitude(nmea_sentence_bytes), 6)
                        longitude = round(self.parse_longitude(nmea_sentence_bytes), 6)
                        day, month, year = self.parse_date(nmea_sentence_bytes)
                        date_string = "{}/{}/{}".format(day, month, year)

                        hour, minute, second = self.parse_time(nmea_sentence_bytes)
                        time_string = "{:02d}:{:02d}:{:02d}".format(hour, minute, second)

                        # Write data to CSV file.
                        writer.writerow({'Date': date_string, 'Time': time_string, 'Latitude': latitude, 'Longitude': longitude})

                        # Print parsed data.
                        print("Date:", date_string, end=', ')
                        print("Time:", time_string, end=', ')
                        print("Latitude:", latitude, end=', ')
                        print("Longitude:" , longitude)
                    except ValueError as e:
                        print("Error:", e)

    def convertTo2DPlane(lat, lon, reference_lat, reference_lon):
        """
        Converts latitude and longitude coordinates to 2D plane coordinates (x, y)
        with respect to a reference latitude and longitude.

        Parameters:
            lat (float): Latitude of the point to convert.
            lon (float): Longitude of the point to convert.
            reference_lat (float): Reference latitude.
            reference_lon (float): Reference longitude.

        Returns:
            tuple: A tuple containing the x and y coordinates in meters, distance in meters and the bearing in degrees.
        """

        # Earth radius in meters.
        EARTH_RADIUS = 6371000.0

        # Convert degrees to radians.
        reference_lat = math.radians(reference_lat)
        reference_lon = math.radians(reference_lon)
        lat = math.radians(lat)
        lon = math.radians(lon)

        # Calculate differences in coordinates.
        deltaLat = lat - reference_lat
        deltaLon = lon - reference_lon

        # Calculate distance between the two points.
        a = math.sin(deltaLat / 2)**2 + math.cos(reference_lat) * math.cos(lat) * math.sin(deltaLon / 2)**2
        c = 2 * math.atan2(math.sqrt(a), math.sqrt(1 - a))
        distance = EARTH_RADIUS * c

        # Calculate bearing.
        angle = math.atan2(math.sin(deltaLon) * math.cos(lat), math.cos(reference_lat) * math.sin(lat) - math.sin(reference_lat) * math.cos(lat) * math.cos(deltaLon))
        bearing = math.degrees(angle)

        # Convert distance and bearing to x-y coordinates.
        x = distance * math.cos(angle)
        y = distance * math.sin(angle)

        return x, y, distance, bearing

File: MATLAB/Lake/GPS.m
% /////////////////////////////////////////////////////////////////////////
% Analyse collected data.
% /////////////////////////////////////////////////////////////////////////


clc; clear all; close all;


% /////////////////////////////////////////////////////////////////////////
% Load data.
% /////////////////////////////////////////////////////////////////////////
boat_data1 = readmatrix('boat_data1.csv'); 
boat_data2 = readmatrix('boat_data2.csv');
sine = readmatrix('sine.csv'); 
walk_home = readmatrix('walk_home.csv'); 
LakeBoundary = readmatrix('LakeBoundary.xlsx'); 
% /////////////////////////////////////////////////////////////////////////


% /////////////////////////////////////////////////////////////////////////
% Analyse data for lake.
% /////////////////////////////////////////////////////////////////////////
latitude1 = boat_data1(:, 3);
longitude1 = boat_data1(:, 4);
bearing1 = boat_data1(:, 5);

% Initialize arrays to store converted coordinates.
x = zeros(size(latitude1));
y = zeros(size(longitude1));

% Convert latitude and longitude to meters.
for i = 2:length(latitude1)
    [x(i), y(i), ~] = convertTo2DPlane(latitude1(i), longitude1(i), latitude1(1), longitude1(1));
end

figure(1);

% Plot 2D Plot of Latitude and Longitude.
subplot(2,1,1);
plot(longitude1, latitude1, '-r');
pbaspect([1 1 1]);
xlabel('Longitude');
ylabel('Latitude');
title('2D Plot of Latitude and Longitude');

% Plot 2D Plot of x and y (in meters).
subplot(2,1,2);
plot(y, x, '-r');
pbaspect([1 1 1]);
xlabel('Y (meters)');
ylabel('X (meters)');
title('2D Plot of x and y (in meters)');

figure(2);
x = 1:length(bearing1);
plot(x, bearing1, '-r');
xlabel('Number of Samples');
ylabel('Heading (Degree)');
title('Heading (Degree)');
% /////////////////////////////////////////////////////////////////////////


% /////////////////////////////////////////////////////////////////////////
% Analyse data for lake.
% /////////////////////////////////////////////////////////////////////////
latitude2 = boat_data2(:, 3);
longitude2 = boat_data2(:, 4);
bearing2 = boat_data2(:, 5);

% Initialize arrays to store converted coordinates.
x = zeros(size(latitude2));
y = zeros(size(longitude2));

% Convert latitude and longitude to meters.
for i = 2:length(latitude2)
    [x(i), y(i), ~] = convertTo2DPlane(latitude2(i), longitude2(i), latitude2(1), longitude2(1));
end

figure(3);

% Plot 2D Plot of Latitude and Longitude.
subplot(2,1,1);
plot(longitude2, latitude2, '-r');
pbaspect([1 1 1]);
xlabel('Longitude');
ylabel('Latitude');
title('2D Plot of Latitude and Longitude');

% Plot 2D Plot of x and y (in meters).
subplot(2,1,2);
plot(y, x, '-r');
pbaspect([1 1 1]);
xlabel('Y (meters)');
ylabel('X (meters)');
title('2D Plot of x and y (in meters)');

figure(4);
x = 1:length(bearing2);
plot(x, bearing2, '-r');
xlabel('Number of Samples');
ylabel('Heading (Degree)');
title('Heading (Degree)');
% /////////////////////////////////////////////////////////////////////////


% /////////////////////////////////////////////////////////////////////////
% Analyse data for lake.
% /////////////////////////////////////////////////////////////////////////
latitude3 = sine(:, 3);
longitude3 = sine(:, 4);
bearing3 = sine(:, 5);

% Initialize arrays to store converted coordinates.
x = zeros(size(latitude3));
y = zeros(size(longitude3));

% Convert latitude and longitude to meters.
for i = 2:length(latitude3)
    [x(i), y(i), ~] = convertTo2DPlane(latitude3(i), longitude3(i), latitude3(1), longitude3(1));
end

figure(4);

% Plot 2D Plot of Latitude and Longitude.
subplot(2,1,1);
plot(longitude3, latitude3, '-r');
pbaspect([1 1 1]);
xlabel('Longitude');
ylabel('Latitude');
title('2D Plot of Latitude and Longitude');

% Plot 2D Plot of x and y (in meters).
subplot(2,1,2);
plot(y, x, '-r');
pbaspect([1 1 1]);
xlabel('Y (meters)');
ylabel('X (meters)');
title('2D Plot of x and y (in meters)');

figure(5);
x = 1:length(bearing3);
plot(x, bearing3, '-r');
xlabel('Number of Samples');
ylabel('Heading (Degree)');
title('Heading (Degree)');
% /////////////////////////////////////////////////////////////////////////


% /////////////////////////////////////////////////////////////////////////
% Comparison.
% /////////////////////////////////////////////////////////////////////////
figure(6);
plot(longitude1, latitude1, '-r', longitude2, latitude2, '-b', longitude3, latitude3, '-g');
pbaspect([1 1 1]);
legend('Reading 1', 'Reading 2', 'Reading 3')
xlabel('Longitude');
ylabel('Latitude');
title('Comparison for 2D Plot of Latitude and Longitude');
% /////////////////////////////////////////////////////////////////////////


% /////////////////////////////////////////////////////////////////////////
% Lake boundary.
% /////////////////////////////////////////////////////////////////////////
figure(7);
lat = boat_data1(:, 3);
lon = boat_data1(:, 4);

plot(lon, lat, '-r');
pbaspect([1 1 1]);
xlabel('Longitude');
ylabel('Latitude');
title('2D Plot of Latitude and Longitude');
% /////////////////////////////////////////////////////////////////////////


function [x, y, bearing, distance] = convertTo2DPlane(lat2, lon2, reference_lat, reference_lon)
    % Earth radius in meters.
    EARTH_RADIUS = 6371000.0;

    % Convert degrees to radians.
    reference_lat = deg2rad(reference_lat);
    reference_lon = deg2rad(reference_lon);
    lat2 = deg2rad(lat2);
    lon2 = deg2rad(lon2);

    % Calculate differences in coordinates.
    deltaLat = lat2 - reference_lat;
    deltaLon = lon2 - reference_lon;

    % Calculate distance between the two points.
    a = sin(deltaLat / 2)^2 + cos(reference_lat) * cos(lat2) * sin(deltaLon / 2)^2;
    c = 2 * atan2(sqrt(a), sqrt(1 - a));
    distance = EARTH_RADIUS * c;

    % Calculate bearing.
    angle = atan2(sin(deltaLon) * cos(lat2), cos(reference_lat) * sin(lat2) - sin(reference_lat) * cos(lat2) * cos(deltaLon));
    bearing = rad2deg(angle);

    % Convert distance and bearing to x-y coordinates.
    x = distance * cos(angle);
    y = distance * sin(angle);
end